### ISSUE TITLE
1.10.2 major fps drop from Forge 2025 onwards

### ISSUE THREAD

**Issue by Zygus42** [williewillus/Botania#462]
I've been tracking a major fps drop on our 1.10.2 build apparently being caused by Forge - this is only significant on our older machines with AMD video cards, as they're already running very fps marginally.

Tracked it to version 2025, but it didn't do it on a non-modded build.  So I tracked down the mod, which ended up being Botania (Botania-unofficial.r1.8-312). 

According to the forge changelogs:
        Build 2025:
            mezz: Force mipmap even if textures are the wrong size

And in the logs for 2024 (but not for 2025):
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara8.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara9.png with size 60x60 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara6.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara7.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara4.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara5.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara2.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara3.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.
[09:17:44] [Client thread/WARN]: Texture botania:textures/items/flightTiara1.png with size 20x20 will have visual artifacts at mip level 4, it can only support level 2. Please report to the mod author that the texture should be some multiple of 16x16.

**Comment by williewillus** [comment]
Hm I know the textures are missized but this shouldn't be the main contributor of the issue. It doesn't happen when you move botania or turn off mipmapping?

**Comment by williewillus** [comment]
(or turning it down to 2 like the log suggests)

**Comment by Zygus42** [comment]
Doesn't do it without Botania.  Doesn't do it prior to forge 2025.

Will try turning mip mapping down and see what happens - yep, dropping it down to 2 fixes the problem.  Don't think I really want to do that for 1 item though, but is a workaround for now I guess.

**Comment by williewillus** [comment]
Thanks for the info, I will fix the texture.

But @mezz should these drops in performance be happening?

**Comment by williewillus** [comment]
https://github.com/Vazkii/BotaniaV2/issues/1

**Comment by mezz** [comment]
> "Why do my eyes hurt?"
> "Because you've never used them before..."

The lower performance is because mipmapping level was always lowered or disabled before, due to mods using oddly sized textures. Now it's always enabled as it should be, which uses more ram on your graphics card. You can set the mipmap level lower to make things like it was before.

**Comment by MaPePeR** [comment]
That's a weird explanation for someone with an uninformed standpoint,  like me,  @mezz

"The lower performance is because it was lowered or disabled before. Now it is always enabled. You can change the setting to increase the performance. "

Isn't mipmapping supposed to increase performance? Or is this a hardware specific issue? If it is  the default value needs to be chosen in a smarter way dependent on the hardware...

**Comment by mezz** [comment]
Sorry, I'll try to explain better. 

Mipmapping is a graphics enhancement, it helps make distant textures look better. It requires more graphics memory though, so if you are almost maxing out your card's memory then adding mipmapping can push it over the limit and kill performance. 

Modded adds lots of textures, but vanilla Minecraft does not have many textures and has set mipmapping to the max value by default since even old cards can handle it. There is no "smart" detection going on. The right value is dependent on the card's ram, the mods installed, and the texture pack used (32x or 64x textures will use much more graphics memory).

Try setting mipmapping to 0, it should fix your performance problem. Earlier, Zygus42 stated that setting it down to 2 fixed it for them.

I explain the reasoning behind what I changed and why here: 
https://github.com/MinecraftForge/MinecraftForge/pull/3082
and in the updated version of the PR which was merged:
https://github.com/MinecraftForge/MinecraftForge/pull/3087
